CMakeLists.txt

```diff
@@ -633,24 +583,133 @@ if (WIN32)
+if (ANDROID)
+	target_link_libraries(torrent-rasterbar PRIVATE ${CMAKE_DL_LIBS})
 endif()
 
-if(APPLE)
+if (APPLE)
 	# for ip_notifier
 	target_link_libraries(torrent-rasterbar PRIVATE "-framework CoreFoundation" "-framework SystemConfiguration")
-endif (APPLE)
+endif()
+
+feature_option(build_tests "build tests" OFF)
+feature_option(build_examples "build examples" OFF)
+feature_option(build_tools "build tools" OFF)
+feature_option(python-bindings "build python bindings" OFF)
+
+# these options require existing target
+feature_option(dht "enable support for Mainline DHT" ON)
+if(NOT build_tests) # tests require deprecated symbols
+	target_optional_compile_definitions(torrent-rasterbar PUBLIC FEATURE NAME deprecated-functions DEFAULT ON
+		DESCRIPTION "enable deprecated functions for backwards compatibility" DISABLED TORRENT_NO_DEPRECATE)
+endif()
+feature_option(encryption "Enables encryption in libtorrent" ON)
+feature_option(exceptions "build with exception support" ON)
+target_optional_compile_definitions(torrent-rasterbar PUBLIC FEATURE NAME extensions DEFAULT ON
+	DESCRIPTION "Enables protocol extensions" DISABLED TORRENT_DISABLE_EXTENSIONS
+)
+target_optional_compile_definitions(torrent-rasterbar PUBLIC FEATURE NAME extensions DEFAULT ON
+	DESCRIPTION "Enables protocol extensions" DISABLED TORRENT_DISABLE_EXTENSIONS)
+target_optional_compile_definitions(torrent-rasterbar PUBLIC FEATURE NAME i2p DEFAULT ON
+	DESCRIPTION "build with I2P support" DISABLED TORRENT_USE_I2P=0)
+target_optional_compile_definitions(torrent-rasterbar PUBLIC FEATURE NAME ipv6 DEFAULT ON
+	DESCRIPTION "build with IPv6 support" DISABLED TORRENT_USE_IPV6=0)
+target_optional_compile_definitions(torrent-rasterbar PUBLIC FEATURE NAME logging DEFAULT ON
+	DESCRIPTION "build with logging" DISABLED TORRENT_DISABLE_LOGGING)
+
+find_public_dependency(Iconv)
+set_package_properties(Iconv
+	PROPERTIES
+		URL "https://www.gnu.org/software/libiconv/"
+		DESCRIPTION "GNU encoding conversion library"
+		TYPE RECOMMENDED
+		PURPOSE "Convert strings between various encodings"
+)
+
+if (Iconv_FOUND)
+	target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_ICONV)
+	target_include_directories(torrent-rasterbar PRIVATE ${ICONV_INCLUDE_DIR})
+	target_link_libraries(torrent-rasterbar PRIVATE ${ICONV_LIBRARIES})
+endif()
+
+find_public_dependency(OpenSSL)
+set_package_properties(OpenSSL
+	PROPERTIES
+		URL "https://www.openssl.org/"
+		DESCRIPTION "Full-strength general purpose cryptography library"
+		TYPE RECOMMENDED
+		PURPOSE "Provides HTTPS support to libtorrent"
+)
 
-if (NOT deprecated-functions)
-	target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_NO_DEPRECATE)
+if(OPENSSL_FOUND)
+	target_link_libraries(torrent-rasterbar PUBLIC OpenSSL::SSL)
+	target_compile_definitions(torrent-rasterbar PUBLIC TORRENT_USE_OPENSSL)
+	target_sources(torrent-rasterbar PRIVATE src/pe_crypto)
+else()
+	if (NOT WIN32 AND NOT APPLE)
+		target_sources(torrent-rasterbar PRIVATE src/sha1)
```

Comment:
libtorrent can also use libgcrypt for sha1 so the internal implementation should only be linked in if neither openssl nor libgcrypt are being used.